File: NET/JNetCall.Sharp.Common/Impl/IO/StreamTransport.cs
using System;
using System.IO;
using JNetBase.Sharp.IO;
using JNetCall.Sharp.API.Enc;
using JNetCall.Sharp.API.IO;

namespace JNetCall.Sharp.Impl.IO
{
    public sealed class StreamTransport : ISendTransport, IPullTransport, IDisposable
    {
        private readonly IEncoding<byte[]> _encoding;
        private readonly Stream _streamIn;
        private readonly Stream _streamOut;

        public StreamTransport(IEncoding<byte[]> encoding,
            Stream streamIn, Stream streamOut)
        {
            _encoding = encoding;
            _streamIn = streamIn;
            _streamOut = streamOut;
        }

        public void Send<T>(T payload)
        {
            var bytes = _encoding.Encode(payload);
            _streamOut.Write(bytes, 0, bytes.Length);
            _streamOut.Flush();
            if (_streamOut is IRewindable r)
                r.Rewind(bytes.Length);
        }

        public T Pull<T>()
        {
            var bytes = StreamTools.ReadWithSize(_streamIn);
            return _encoding.Decode<T>(bytes);
        }

        public void Dispose()
        {
            _streamOut.Dispose();
            _streamIn.Dispose();
            _encoding.Dispose();
        }
    }
}

File: JVM/jnetcall-java-common/src/main/java/jnetcall/java/impl/io/StreamTransport.java
package jnetcall.java.impl.io;

import java.io.InputStream;
import java.io.OutputStream;

import jnetbase.java.io.IRewindable;
import jnetcall.java.api.enc.IEncoding;
import jnetcall.java.api.io.IPullTransport;
import jnetcall.java.api.io.ISendTransport;

public final class StreamTransport implements ISendTransport, IPullTransport, AutoCloseable {

    private final IEncoding<byte[]> _encoding;
    private final InputStream _streamIn;
    private final OutputStream _streamOut;

    public StreamTransport(IEncoding<byte[]> encoding,
                            InputStream streamIn, OutputStream streamOut) {
        _encoding = encoding;
        _streamIn = streamIn;
        _streamOut = streamOut;
    }

    @Override
    public <T> T pull(Class<T> clazz) {
        try {
            byte[] bytes = StreamTools.readWithSize(_streamIn);
            return _encoding.decode(bytes, clazz);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public <T> void send(T payload) {
        try {
            byte[] bytes = _encoding.encode(payload);
            _streamOut.write(bytes, 0, bytes.length);
            _streamOut.flush();
            if (_streamOut instanceof IRewindable) {
            	IRewindable r = (IRewindable)_streamOut;
                r.rewind(bytes.length);
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void close() throws Exception {
        _streamOut.close();
        _streamIn.close();
        _encoding.close();
    }
}
